### PR TITLE
Enable defining nested data types

### PR DESCRIPTION
**Pull request by PhilippeMoussalli**
PR that addresses #178 
Inspiration: https://swagger.io/docs/specification/data-models/data-types/#:~:text=the%20null%20value.-,Arrays,-Arrays%20are%20defined (credit to @GeorgesLorre)

It includes: 
* Changing the common json schema to enable defining array types 
* Changing the `Type` class to be dynamic rather than typed to enable defining nested structures without explicitly defining them
* Adding an addition test file for `schema.py`
* Updating the current existing examples (embedding + segmentation)

### CODE REVIEW

File: components/segment_images/fondant_component.yaml [hunk 1]
```diff
@@ -12,7 +12,9 @@ produces:
   segmentations:
     fields:
       data:
-        type: binary
```
Comment (NielsRogge) [inline components/segment_images/fondant_component.yaml:15]: Same question here

File: components/embedding_based_laion_retrieval/fondant_component.yaml [hunk 2]
```diff
@@ -6,7 +6,9 @@ consumes:
   embeddings:
     fields:
       data:
-        type: float32_list
+        type: array
```
[the remark this answers is not in the document]
Comment (PhilippeMoussalli) [reply components/embedding_based_laion_retrieval/fondant_component.yaml:9]: it's represented as a list/array of floats 

`pa._list(pa.float32()) == List[float] ==  {"type": "array", "items": {"type": "float32"}}`
(1st is pyarrow, 2nd is python data type, third is json schema)
https://swagger.io/docs/specification/data-models/data-types/#:~:text=the%20null%20value.-,Arrays,-Arrays%20are%20defined

[the remark this answers is not in the document]
Comment (NielsRogge) [reply components/embedding_based_laion_retrieval/fondant_component.yaml:9]: Ok makes sense!

File: fondant/schema.py [hunk 3]
```diff
@@ -2,56 +2,148 @@
...
+        """
+        Validates the provided data type and returns the corresponding data type object.
+
+        Args:
+            data_type: The data type to validate.
+
+        Returns:
+            The validated `pa.DataType` object.
+        """
+        if isinstance(data_type, str):
```
Comment (RobbeSneyders) [inline fondant/schema.py:67]:
```suggestion
        if not isinstance(data_type, Type):
```

I think this is a bit more robust. What if I pass in something that is not a `Type` nor a string? :)

File: fondant/schema.py [hunk 4]
```diff
@@ -2,56 +2,148 @@
...
+        Args:
+            json_schema: The dictionary representation of the data type, can represent nested values
+
+        Returns:
+            A new `Type` instance representing the specified data type.
+        """
+        if json_schema["type"] == "array":
+            items = json_schema["items"]
+            if isinstance(items, dict):
+                return cls.list(Type.from_json(items))
```
Comment (RobbeSneyders) [inline fondant/schema.py:109]:
```suggestion
                return cls.list(cls.from_json(items))
```

File: fondant/schema.py [hunk 5]
```diff
@@ -2,56 +2,148 @@
...
+        else:
+            return Type(json_schema["type"])
```
Comment (RobbeSneyders) [inline fondant/schema.py:111]:
```suggestion
            return cls(json_schema["type"])
```

File: tests/test_manifest.py [hunk 6]
```diff
@@ -60,16 +60,16 @@ def test_subset_fields():
     subset = Subset(specification=subset_spec, base_path="/tmp")
 
     # add a field
-    subset.add_field(name="data2", type_=Type.binary)
+    subset.add_field(name="data2", type_=Type("binary"))
```
Comment (RobbeSneyders) [inline tests/test_manifest.py:63]: It's a bit unfortunate that we lose the ability to reference types by attribute (and therefore autocomplete, static analysis, etc., ...). Unfortunately I don't see a straightforward way to keep them either.

**Review by RobbeSneyders — COMMENTED** [review]
Thanks @PhilippeMoussalli, left some small comments.

File: tests/test_manifest.py (hunk 6 above)
Comment (PhilippeMoussalli) [reply tests/test_manifest.py:63]: yeah indeed, you would need to know/type all possible combinations which is not very feasible. But it's not user-facing so it should not have that big of an impact.

**Review by RobbeSneyders — APPROVED** [review]

**Pull request merged**